crates/matrix-sdk/src/widget/matrix.rs

```diff
@@ -330,38 +340,171 @@ impl MatrixDriver {
                 | "m.secret.send"
                 // drop utd traffic
                 | "m.room.encrypted"
-        );
-
-        if filtered {
-            trace!("To-device message of type <{event_type}> filtered out by widget driver.",);
-        }
-        filtered
+        )
     }
 
-    /// It will ignore all devices where errors occurred or where the device is
-    /// not verified or where th user has a has_verification_violation.
+    /// If the room the widget is in is encrypted, then the to-device message
+    /// will be encrypted. If one of the named devices does not exist, then
+    /// the call will fail with an error.
     pub(crate) async fn send_to_device(
         &self,
         event_type: ToDeviceEventType,
-        encrypted: bool,
         messages: BTreeMap<
             OwnedUserId,
             BTreeMap<DeviceIdOrAllDevices, Raw<AnyToDeviceEventContent>>,
         >,
-    ) -> Result<send_event_to_device::v3::Response> {
+    ) -> Result<SendToDeviceEventResponse> {
+        // TODO: block this at the negotiation stage, no reason to let widget believe
+        // they can do that
+        if Self::is_internal_type(&event_type.to_string()) {
+            warn!("Widget tried to send internal to-device message <{}>, ignoring", event_type);
+            // Silently return a success response, the widget will not receive the message
+            return Ok(Default::default());
+        }
+
         let client = self.room.client();
 
-        let request = if encrypted {
-            return Err(Error::UnknownError(
-                "Sending encrypted to-device events is not supported by the widget driver.".into(),
-            ));
+        let mut failures: BTreeMap<OwnedUserId, Vec<OwnedDeviceId>> = BTreeMap::new();
```

Comment:
Is there a point to initializing this here, mutably borrowing it in the `encrypt_and_send_content_to_devices_helper` call?

Can't `encrypt_and_send_content_to_devices_helper` just return this?

Reply:
Even if I return a failure map from `encrypt_and_send_content_to_devices_helper` I will still have to intialize a map outside of the loop (`for .. in content_to_recipients_map` )in order to accumalate all the failures. Not sure if it really makes a differenc